shm/src/ship.cpp

```diff
@@ -0,0 +1,95 @@
+#include "ship.hpp"
+
+#include <algorithm>
+#include <numeric>
+
+Ship::Ship(int maxCrew, int speed, size_t id, Time* time, Delegate* delegate)
+    : Ship(0, maxCrew, speed, "", id, time, delegate)
+{}
+Ship::Ship(int capacity, int maxCrew, int speed,
+        const std::string& name, size_t id, Time* time, Delegate* delegate)
+    : capacity_(capacity)
+    , maxCrew_(maxCrew) 
+    , crew_(0)
+    , speed_(speed)
+    , name_(name)
+    , id_(id)
+    , time_(time)
+    , delegate_(delegate)
+{
+    time_->addObserver(this);
+}
+Ship::~Ship() {
+    time_->removeObserver(this);
+}
+
+Ship& Ship::operator-=(const size_t crew) {
+    if (crew <= crew_) {
+        crew_ -= crew;
+        return *this;
+    }
+    return *this;
+}
+Ship& Ship::operator+=(const size_t crew) {
+    if (crew_ + crew <= maxCrew_) {

```

Comment:
I think it's better to compare crew_ and crew, because if we add crew_ and crew we risk going out of size_t range,which may result in gargabe value that can be lower than crew limit.